flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -969,25 +969,25 @@ private void dropSubsumedCheckpoints(long checkpointId) {
 	}
 
 	/**
-	 * Triggers the queued request, if there is one.
+	 * Resumes suspended periodic triggering.
 	 *
 	 * <p>NOTE: The caller of this method must hold the lock when invoking the method!
 	 */
-	private void triggerQueuedRequests() {
-		if (triggerRequestQueued) {
-			triggerRequestQueued = false;
+	private void resumePeriodicTriggering() {
+		assert(Thread.holdsLock(lock));
+
+		if (shutdown || !periodicScheduling) {
+			return;
+		}
+		if (periodicTriggeringSuspended) {
+			periodicTriggeringSuspended = false;
 
 			// trigger the checkpoint from the trigger timer, to finish the work of this thread before
 			// starting with the next checkpoint
-			if (periodicScheduling) {
-				if (currentPeriodicTrigger != null) {
-					currentPeriodicTrigger.cancel(false);
-				}
-				currentPeriodicTrigger = scheduleTriggerWithDelay(0L);
-			}
-			else {
-				timer.execute(new ScheduledTrigger());
```

Comment:
Why has this behaviour disappeared?

Reply:
Actually there is a difference between `triggerQueuedRequests` and `resumePeriodicTriggering`. It's about the non-periodic checkpoint triggering (not savepoint). However there is no such a scenario (non-periodic checkpoint) in production codes. It's dead code somewhat.

This part is about the non-periodic checkpoint. In the `triggerQueuedRequests` version, it would re-schedule a one-shot checkpoint trigger if `periodicScheduling` is false. I guess the reason is that it assume the `triggerRequestQueued` is triggered by a non-periodic checkpoint trigger. So it recovers a non-periodic trigger. While in the `resumePeriodicTriggering` version, it would do nothing if periodic triggering is disabled.

Let's assume there is a non-periodic trigger scenario. Re-scheduling an one-shot trigger also makes no sense to me. I don't think this behavior worth being kept. If caller wants an one-shot checkpoint being executed anyway, the `forceCheckpoint` should be set. Otherwise this one-shot checkpoint might fail and will never be re-scheduled. Caller could track the lifecycle thought `CompletableFuture` returned by `triggerCheckpoint`.

Reply:
I guess your concern is that some necessary trigger might be missing?
Stopping with savepoint would cancel all in-flight checkpoints/savepoints. And it does not allow any further periodic triggering. The `ScheduledTrigger` is just a normal one-shot checkpoint. I think it's fine to remove it. I didn't find any reason to keep it. I believe the reason it stays here is the unclear semantics of `triggerQueuedRequests`. There is a trigger request "queued", we should "dequeue" something afterwards.